File: walker_ai/src/reset_controls.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_services_default

from controller_manager_msgs.srv import SwitchController
from gazebo_msgs.srv import SpawnEntity, DeleteEntity
from geometry_msgs.msg import Pose
from builtin_interfaces.msg import Duration

class ControllerResetClient(Node):
    def __init__(self):
        super().__init__("controller_reset")
        self.get_logger().info("creating client...")
        self.cli = self.create_client(SwitchController, "/controller_manager/switch_controller", qos_profile=qos_profile_services_default)
        while not self.cli.wait_for_service():
            self.get_logger().info("service not available, waiting again...")
        self.req = SwitchController.Request()
    
    def send_request(self, activate_controllers=[], deactivate_controllers=[]):
        self.req.activate_controllers = activate_controllers
        self.req.deactivate_controllers = deactivate_controllers
        self.req.activate_asap = True
        self.req.strictness=2 #BEST_EFFORT=1, STRICT=2
        self.future = self.cli.call_async(self.req)
        self.req.timeout=Duration()
        self.req.timeout.sec = 10
        rclpy.spin_until_future_complete(self, self.future)
        return self.future.result()

    def reset_controls(self):
        #reset controls from deactivation to activation
        self.get_logger().info("resetting controls...")
        result = False
        controller_arr = ["joint_effort_controller", "joint_state_broadcaster"]
        deactivate_ok = self.send_request(activate_controllers=[], deactivate_controllers=controller_arr)
        if deactivate_ok:
            self.get_logger().info("deactivated the controller")
            activate_ok = self.send_request(activate_controllers=controller_arr, deactivate_controllers=[])
            if activate_ok:
                self.get_logger().info("controllers reset!")
                result = True
            else:
                self.get_logger().info("activate_ok ==> " + str(activate_ok))
        else:
            self.get_logger().info("DEactivate_ok ==> " + str(deactivate_ok))

        return result
    

if __name__ == "__main__":
    rclpy.init()
    reset_controls = ControllerResetClient()
    reset_controls.get_logger().info("client created!")
    response = reset_controls.reset_controls()
    reset_controls.get_logger().info("controllers reset!")
    reset_controls.get_logger().info("here's strictness: "+str(reset_controls.req.strictness))
    reset_controls.destroy_node()
    rclpy.shutdown()

File: walker_ai/src/reset_controls.cpp
#include "rclcpp/rclcpp.hpp"
#include "controller_manager_msgs/srv/switch_controller.hpp"

#include <memory>
#include <string>
#include <vector>

using namespace std::chrono_literals;


std::vector<std::string> controller_arr = {"joint_effort_controller", "joint_state_broadcaster"};

class ControllerResetClient : public rclcpp::Node{
    public:
        //member vars
        rclcpp::Client<controller_manager_msgs::srv::SwitchController>::SharedPtr cli;
        std::shared_ptr<controller_manager_msgs::srv::SwitchController::Request> request;

        ControllerResetClient() : Node("controller_reset_cpp"){
            cli = this->create_client<controller_manager_msgs::srv::SwitchController>("/controller_manager/switch_controller");
            while (!cli->wait_for_service(1s)) {
                if (!rclcpp::ok()) {
                    RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
                }
                RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
            }
            request = std::make_shared<controller_manager_msgs::srv::SwitchController::Request>();
        }
        bool send_request(std::vector<std::string> activate_controllers, std::vector<std::string> deactivate_controllers){
            request->activate_controllers = activate_controllers;
            request->deactivate_controllers = deactivate_controllers;
            auto future = cli->async_send_request(request);
            if (rclcpp::spin_until_future_complete(this->get_node_base_interface(), future) == rclcpp::FutureReturnCode::SUCCESS) return true;
            else return false;
        }

        bool reset_controls(){
            RCLCPP_INFO(this->get_logger(), "restetting controls");
            bool result = false;
            bool deactivateOk = send_request({}, controller_arr);
            if (deactivateOk){
                bool activateOk = send_request(controller_arr, {});
                if (activateOk){
                    result = true;
                    RCLCPP_INFO(this->get_logger(), "activate successful!");
                }
                else RCLCPP_INFO(this->get_logger(), "activate failed");
            }
            else{
                RCLCPP_INFO(this->get_logger(), "ERROR: deactivate_ok is False!");
            }
            return result;
        }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    std::shared_ptr<ControllerResetClient> node = std::make_shared<ControllerResetClient>();
    rclcpp::shutdown();
    return 0;
}
